src/async_impl/decoder.rs

```diff
@@ -0,0 +1,297 @@
+/*!
+A potentially non-blocking response decoder.
+
+The decoder wraps a stream of chunks in a `Read` interface. 
+If the underlying chunk stream never blocks (by returning `Err(WouldBlock)`),
+then the decoder will also never return `Err(WouldBlock)`.
+This is essential for it to be usable in synchronous scenarios.
+
+If the response is potentially gzipped, then we need to read a byte to make sure the response isn't actually empty.
+This is also done in an asynchronous in case the response body isn't immediately available.
+
+This module consists of a few main types:
+
+- `ReadableChunks` is a `Read`-like wrapper around a stream
+- `PollChunk` is a stream of chunks, that may or may not be non-blocking
+- `Decoder` is a layer over `ReadableChunks` that applies the right decompression
+
+The following types directly support the gzip compression case:
+
+- `DecoderInnerFuture` is a non-blocking constructor for a `Decoder` in case the body needs to be read
+- `Peeked` is a wrapper around a peeked byte that prevents an EOF on read
```

Comment:
You could imagine a less-stellar server setting a header universally, but then it turns out the file it is serving is empty.

Of course, in this case, since the bytes are coming in a `Stream` of `Chunks`, we can detect EOF before ever using the libflate encoder: if `poll` returns `Ok(Async::Ready(None))`, that's the EOF. No need to ask libflate to decode then :)